src/action/base/create_or_merge_nix_config.rs

```diff
@@ -0,0 +1,490 @@
+            mode = tracing::field::display(format!("{:#o}", NIX_CONF_MODE)),
+            merged_nix_config = tracing::field::Empty,
+        );
+
+        if tracing::enabled!(tracing::Level::TRACE) {
+            span.record(
+                "merged_nix_config",
+                &self
+                    .nix_configs
+                    .merged_nix_config
+                    .settings()
+                    .iter()
+                    .map(|(k, v)| format!("{k}='{v}'"))
+                    .collect::<Vec<_>>()
+                    .join(" "),
+            );
+        }
+        span
+    }
+
+    fn execute_description(&self) -> Vec<ActionDescription> {
+        vec![ActionDescription::new(
+            self.tracing_synopsis(),
+            vec![format!(
+                "If {} already exists, we will attempt to merge the current settings with our settings; \
+                otherwise, it will be created with only our settings",
+                self.path.display()
+            )],
+        )]
+    }
+
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn execute(&mut self) -> Result<(), ActionError> {
+        let Self { path, nix_configs } = self;
+
+        if tracing::enabled!(tracing::Level::TRACE) {
+            let span = tracing::Span::current();
+            span.record(
+                "merged_nix_config",
+                nix_configs
+                    .merged_nix_config
+                    .settings()
+                    .iter()
+                    .map(|(k, v)| format!("{k}='{v}'"))
+                    .collect::<Vec<_>>()
+                    .join(" "),
+            );
+        }
+
+        let mut options = OpenOptions::new();
+        options.create(true).write(true).read(true);
+        options.mode(NIX_CONF_MODE);
+
+        // Create a temporary file in the same directory as the one
+        // that the final file goes in, so that we can rename it
+        // atomically
+        let parent_dir = path.parent().expect("File must be in a directory");
+        let mut temp_file_path = parent_dir.to_owned();
+        {
+            let mut rng = rand::thread_rng();
+            temp_file_path.push(format!("nix-installer-tmp.{}", rng.gen::<u32>()));
+        }
+        let mut temp_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            // If the file is created, ensure that it has harmless
+            // permissions regardless of whether the mode will be
+            // changed later (if we ever create setuid executables,
+            // they should only become setuid once they are owned by
+            // the appropriate user)
+            .mode(0o600)
+            .open(&temp_file_path)
+            .await
+            .map_err(|e| {
+                ActionError::Open(temp_file_path.clone(), e)
+            })?;
+
+        // FIXME(@cole-h): for now we replace the entire file, but in the future we could potentially "replace" the contents
```

Comment:
I wonder if we actually want to do this. It's cheap to replace the whole file, but that does mean the file loses all of its comments (if any).

I think no matter which way we go, though, we should probably not care about comments, because that adds a whole bunch of complexity.

Reply:
I feel like if a user has comments in their config we should make an attempt to preserve them... Could we perhaps get the list of relevant configs to revert and then find their lines within the file and update just the line?

Reply:
I could give it a shot. We already have a list of the configs we're updating as `merged_nix_config`, so we can just iterate over the lines of the file until we find a line that starts with one of those keys and do the fun logic there.

I think I had a WIP of that earlier in this branch, I'll try to find it.

Reply:
Latest commits have resurrected the "preserve comments" stuff (and made it more robust).